File: interop/cpp/plugin_host.cpp
// REPE plugin-ABI interop host.
//
// A C++ REPE host that `dlopen`s a Rust plugin built by the `repe` crate and
// drives it through `glaze/rpc/repe/plugin.h` — the ABI both implementations
// share. Where the fixture generator pins the *wire format* against Glaze, this
// pins the *plugin ABI* against it: symbol names and signatures, the metadata
// struct's layout, the response-buffer contract, and the lifecycle.
//
// Every frame here is encoded by Glaze (`repe::request_json` / `repe::to_buffer`)
// and every response decoded by Glaze (`repe::from_buffer` / `repe::decode_message`).
// That is the point of the file: nothing hand-rolls a header, so a divergence in
// either implementation's framing shows up as a failed decode rather than as a
// byte comparison that both sides got equally wrong.
//
// Build via interop/cpp/CMakeLists.txt; see interop/README.md. Usage:
//
//   plugin_host <path-to-plugin-library>
//
// Exits non-zero on the first failed expectation, with the failures listed.

#include <dlfcn.h>

#include <cstdint>
#include <iostream>
#include <string>
#include <string_view>
#include <vector>

#include "glaze/glaze.hpp"
#include "glaze/rpc/repe/buffer.hpp"
#include "glaze/rpc/repe/plugin.h"
#include "glaze/rpc/repe/repe.hpp"

namespace repe = glz::repe;

namespace
{
   int failures = 0;

   void check(bool ok, std::string_view what)
   {
      std::cout << (ok ? "  ok   " : "  FAIL ") << what << '\n';
      if (!ok) {
         ++failures;
      }
   }

   // The five entry points, resolved by name exactly as a production host does.
   // `init` and `shutdown` are optional per plugin.h ("may be NULL"), so they are
   // resolved but not required.
   struct plugin_abi
   {
      uint32_t (*interface_version)(void){};
      const repe_plugin_data* (*info)(void){};
      repe_result (*init)(void){};
      void (*shutdown)(void){};
      repe_buffer (*call)(const char*, uint64_t){};
   };

   // Call across the ABI and copy immediately. The returned buffer is valid only
   // until this thread's next call, so a host that keeps it must copy first —
   // doing that here is not defensive coding, it is the contract.
   //
   // Returns std::nullopt for a zero-size response, which is what a notify
   // produces and which a host must read as "send nothing" rather than an error.
   std::optional<std::string> host_call(const plugin_abi& abi, std::string_view request)
   {
      const repe_buffer out = abi.call(request.data(), request.size());
      if (out.data == nullptr) {
         check(false, "response data pointer is never null");
         return std::nullopt;
      }
      // Constructing a string_view from (data, 0) is exactly what a C++ host
      // does and is undefined if data is null — the reason the plugin returns a
      // real address even when empty.
      const std::string_view view{out.data, out.size};
      if (view.empty()) {
         return std::nullopt;
      }
      return std::string{view};
   }

   // Decode a response frame with Glaze and pull out the typed body.
   template <auto Opts = glz::opts{}, class T>
   bool decode_body(const std::string& frame, T& value, std::string_view what)
   {
      repe::message msg{};
      if (repe::from_buffer(frame, msg) != glz::error_code::none) {
         check(false, std::string{what} + ": Glaze could not parse the response frame");
         return false;
      }
      if (const auto err = repe::decode_message<Opts>(value, msg)) {
         check(false, std::string{what} + ": " + *err);
         return false;
      }
      return true;
   }

   // Header fields of a response, for the cases where the *error* is the result.
   repe::message parse(const std::string& frame)
   {
      repe::message msg{};
      (void)repe::from_buffer(frame, msg);
      return msg;
   }
}

int main(int argc, char** argv)
{
   if (argc < 2) {
      std::cerr << "usage: plugin_host <path-to-plugin-library>\n";
      return 2;
   }
   const char* path = argv[1];

   void* handle = dlopen(path, RTLD_NOW | RTLD_LOCAL);
   if (!handle) {
      std::cerr << "dlopen(" << path << ") failed: " << dlerror() << '\n';
      return 2;
   }
   // Deliberately never dlclose: the plugin's response buffer is a thread-local,
   // and unloading while threads that touched it are alive leaves TLS destructor
   // addresses dangling. Documented in docs/plugins.md as a host requirement.

   plugin_abi abi{};
   abi.interface_version = reinterpret_cast<uint32_t (*)(void)>(dlsym(handle, "repe_plugin_interface_version"));
   abi.info = reinterpret_cast<const repe_plugin_data* (*)(void)>(dlsym(handle, "repe_plugin_info"));
   abi.init = reinterpret_cast<repe_result (*)(void)>(dlsym(handle, "repe_plugin_init"));
   abi.shutdown = reinterpret_cast<void (*)(void)>(dlsym(handle, "repe_plugin_shutdown"));
   abi.call = reinterpret_cast<repe_buffer (*)(const char*, uint64_t)>(dlsym(handle, "repe_plugin_call"));

   std::cout << "symbols\n";
   check(abi.interface_version != nullptr, "repe_plugin_interface_version resolves");
   check(abi.info != nullptr, "repe_plugin_info resolves");
   check(abi.call != nullptr, "repe_plugin_call resolves");
   if (!abi.interface_version || !abi.info || !abi.call) {
      return 1; // Nothing below can run without these three.
   }

   // --- version handshake, before the metadata struct is read ---------------
   std::cout << "version\n";
   check(abi.interface_version() == REPE_PLUGIN_INTERFACE_VERSION,
         "plugin reports the interface version this host was built against");
   if (abi.interface_version() != REPE_PLUGIN_INTERFACE_VERSION) {
      // Fatal, not merely recorded. `plugin.h` is explicit that the version is
      // checked *before* the struct layout is interpreted: reading a
      // `repe_plugin_data` written by a different version is what the standalone
      // version function exists to prevent, and this file is the reference for
      // the sequence.
      return 1;
   }

   // --- metadata ------------------------------------------------------------
   std::cout << "metadata\n";
   const repe_plugin_data* info = abi.info();
   check(info != nullptr, "repe_plugin_info returns a struct");
   if (!info) {
      return 1;
   }
   const std::string root = info->root_path;
   std::cout << "       name='" << info->name << "' version='" << info->version << "' root='" << root << "'\n";
   check(!root.empty() && root.front() == '/', "root_path is an absolute JSON Pointer prefix");
   check(root.size() == 1 || root.back() != '/', "root_path carries no trailing separator");

   // --- lifecycle -----------------------------------------------------------
   std::cout << "lifecycle\n";
   if (abi.init) {
      check(abi.init() == REPE_OK, "first repe_plugin_init reports REPE_OK");
      check(abi.init() == REPE_ERROR_ALREADY_INITIALIZED, "second repe_plugin_init reports ALREADY_INITIALIZED");
   }

   // --- a field read: empty body means read ---------------------------------
   std::cout << "field read\n";
   {
      const auto request = repe::to_buffer(repe::request_json(repe::user_header{.query = root + "/gain", .id = 1}));

      // A host routes by prefix: this is how it decides the frame is ours.
      check(repe::extract_query(request).starts_with(root), "Glaze's extract_query matches the plugin's root");

      if (const auto response = host_call(abi, request)) {
         double gain = 0;
         if (decode_body(*response, gain, "read /gain")) {
            check(gain == 1.0, "read /gain returns the constructed value");
         }
         check(parse(*response).header.id == 1, "the response echoes the request id");
         check(repe::extract_query(*response) == root + "/gain", "the response echoes the request query");
      }
      else {
         check(false, "a non-notify read produces a response");
      }
   }

   // --- a field write, then a read that observes it -------------------------
   std::cout << "field write\n";
   {
      const auto write = repe::to_buffer(repe::request_json(repe::user_header{.query = root + "/channel", .id = 2}, 6u));
      check(host_call(abi, write).has_value(), "a write is acknowledged");

      const auto read = repe::to_buffer(repe::request_json(repe::user_header{.query = root + "/channel", .id = 3}));
      if (const auto response = host_call(abi, read)) {
         uint32_t channel = 0;
         if (decode_body(*response, channel, "read /channel")) {
            check(channel == 6, "the read observes what the write left behind");
         }
      }
   }

   // --- a method call taking one argument -----------------------------------
   std::cout << "method call\n";
   {
      const auto request =
         repe::to_buffer(repe::request_json(repe::user_header{.query = root + "/calibrate", .id = 4}, 8.0));
      if (const auto response = host_call(abi, request)) {
         double gain = 0;
         if (decode_body(*response, gain, "/calibrate")) {
            check(gain == 4.0, "the method returns its computed result");
         }
      }
   }

   // --- a method returning Err becomes a REPE error frame -------------------
   std::cout << "method error\n";
   {
      const auto request =
         repe::to_buffer(repe::request_json(repe::user_header{.query = root + "/calibrate", .id = 5}, -1.0));
      if (const auto response = host_call(abi, request)) {
         const auto msg = parse(*response);
         check(msg.header.ec != glz::error_code::none, "a handler Err arrives as an error frame");
         check(std::string_view{msg.body}.find("instrument fault") != std::string_view::npos,
               "the error frame carries the handler's message");
      }
   }

   // --- a zero-argument method ----------------------------------------------
   std::cout << "zero-argument method\n";
   {
      const auto request = repe::to_buffer(repe::request_json(repe::user_header{.query = root + "/identify", .id = 6}));
      if (const auto response = host_call(abi, request)) {
         std::string identity;
         if (decode_body(*response, identity, "/identify")) {
            check(identity.starts_with("instrument fw"), "the method result decodes as a string");
         }
      }
   }

   // --- a BEVE typed numeric array field ------------------------------------
   // `#[repe(typed)]` sends the array through the bulk BEVE encoder, so this
   // crosses the boundary as BEVE while everything above is JSON.
   std::cout << "typed numeric field\n";
   {
      const auto request = repe::to_buffer(repe::request_json(repe::user_header{.query = root + "/samples", .id = 7}));
      if (const auto response = host_call(abi, request)) {
         const auto msg = parse(*response);
         check(msg.header.body_format == repe::body_format::BEVE,
               "a #[repe(typed)] field responds with a BEVE body");
         std::vector<double> samples;
         if (decode_body<glz::opts{glz::BEVE}>(*response, samples, "/samples")) {
            check(samples.size() == 8, "Glaze decodes the typed array to its full length");
         }
      }
   }

   // --- notify: no response at all ------------------------------------------
   std::cout << "notify\n";
   {
      const auto request =
         repe::to_buffer(repe::request_json(repe::user_header{.query = root + "/reset", .id = 8, .notify = true}));
      check(!host_call(abi, request).has_value(), "a notify answers with size 0 and a non-null pointer");
   }

   // --- an unknown method under the plugin's root ---------------------------
   std::cout << "unknown method\n";
   {
      const auto request = repe::to_buffer(repe::request_json(repe::user_header{.query = root + "/absent", .id = 9}));
      if (const auto response = host_call(abi, request)) {
         const auto msg = parse(*response);
         check(msg.header.ec == glz::error_code::method_not_found,
               "an unknown method is method_not_found, not a dropped frame");
      }
   }

   // --- a frame Glaze would never emit --------------------------------------
   std::cout << "malformed frame\n";
   {
      const std::string garbage = "not a repe frame at all";
      if (const auto response = host_call(abi, garbage)) {
         const auto msg = parse(*response);
         check(msg.header.ec != glz::error_code::none, "a malformed frame is answered with an error");
         check(msg.header.id == 0, "the id is 0, because it could not be read");
      }
      else {
         check(false, "a malformed frame is answered, not silently dropped");
      }
   }

   // --- shutdown, last, because it is visible to everything above -----------
   std::cout << "shutdown\n";
   if (abi.shutdown) {
      abi.shutdown();
      const auto request = repe::to_buffer(repe::request_json(repe::user_header{.query = root + "/gain", .id = 10}));
      if (const auto response = host_call(abi, request)) {
         check(parse(*response).header.ec != glz::error_code::none,
               "a call after shutdown is refused with an error");
      }
      else {
         check(false, "a call after shutdown is answered, not ignored");
      }
   }

   std::cout << (failures == 0 ? "\nplugin ABI interop: PASS\n" : "\nplugin ABI interop: FAIL\n");
   return failures == 0 ? 0 : 1;
}

File: examples/plugin_host.rs
//! A REPE plugin host: `dlopen` a plugin and drive it over the C ABI.
//!
//! The mirror of [`repe_plugin`](../repe_plugin/index.html). Build it with
//!
//! ```text
//! cargo build --release --features plugin-host --example plugin_host
//! ```
//!
//! and point it at any conforming plugin:
//!
//! ```text
//! target/release/examples/plugin_host target/release/examples/librepe_plugin.so
//! ```
//!
//! The plugin's implementation language does not come into it. This drives the
//! Rust plugin example and a C++ Glaze plugin with the same code and the same
//! expectations, which is what CI uses it for — see `interop/README.md`.
//!
//! What the host itself has to get right is in `repe::plugin::host`: the version
//! handshake before the metadata is read, the optional lifecycle symbols, and
//! copying each response before the plugin's borrow expires. What is left here
//! is the part that is genuinely the application's — deciding which plugin a
//! request belongs to, and what to do with the answer.
//!
//! Beyond the published surface it expects (`gain`, `channel`, `calibrate`,
//! `identify`, `reset`), every check below holds for any conforming plugin,
//! which is what lets one binary drive two implementations. Behavior that is
//! this crate's rather than the protocol's — the response echoing the request
//! query, say, which Glaze's registry does not do — is pinned in
//! `tests/plugin_host.rs` instead, where the plugin under test is known.

use std::process::ExitCode;

use repe::constants::{ErrorCode, QueryFormat};
use repe::message::{Message, MessageView};
use repe::plugin::host::{HostError, LoadOrigin, Plugin};
use repe::server::Router;

/// Failures seen so far. Every check runs, so one bad expectation reports all of
/// them rather than only the first.
struct Checks {
    failures: u32,
}

impl Checks {
    fn check(&mut self, ok: bool, what: &str) {
        println!("  {} {what}", if ok { "ok  " } else { "FAIL" });
        if !ok {
            self.failures += 1;
        }
    }
}

/// A read: a request with no body. REPE separates a read from a write at the
/// frame level, by whether a body is present, so nothing about the path says
/// which one this is.
fn read(query: &str, id: u64) -> Vec<u8> {
    Message::builder()
        .id(id)
        .query_str(query)
        .query_format(QueryFormat::JsonPointer)
        .build()
        .to_vec()
}

/// A write, or a method call with an argument. Same frame either way: what
/// distinguishes them is what the path names.
fn write<T: serde::Serialize>(query: &str, id: u64, value: &T) -> Vec<u8> {
    Message::builder()
        .id(id)
        .query_str(query)
        .query_format(QueryFormat::JsonPointer)
        .body_json(value)
                .build()
        .to_vec()
}

/// A notify: no response is produced, by protocol.
fn notify(query: &str, id: u64) -> Vec<u8> {
    Message::builder()
        .id(id)
        .notify(true)
        .query_str(query)
        .query_format(QueryFormat::JsonPointer)
        .build()
        .to_vec()
}

fn run(path: &str) -> Result<u32, HostError> {
    // SAFETY: loading a native library runs its initializers. This example is
    // handed a plugin path by whoever runs it, which is exactly the trust
    // decision a real host's operator makes when populating a plugin directory.
    let plugin = unsafe { Plugin::load(path) }?;

    let mut checks = Checks { failures: 0 };
    let root = plugin.root_path().to_string();

    println!("metadata");
    println!(
        "       name='{}' version='{}' root='{}' abi={} origin={:?}",
        plugin.name(),
        plugin.version(),
        root,
        plugin.interface_version(),
        plugin.load_origin()
    );
    // `load` refuses anything that fails these, so reaching here is the check.
    // Restated so the output says what was established rather than only what
    // was read.
    checks.check(root.starts_with('/'), "root_path is an absolute prefix");
    checks.check(!plugin.name().is_empty(), "the plugin reports a name");

    // A host routes with `Plugin::claims`. `MessageView` reads the query out of
    // a frame without decoding the body, which is what makes this cheap enough
    // to do on every request.
    println!("routing");
    {
        let frame = read(&format!("{root}/gain"), 1);
        let query = MessageView::from_slice(&frame)
            .expect("a frame this process just built parses")
            .query_str()
            .expect("the query is UTF-8")
            .to_string();
        checks.check(
            plugin.claims(&query),
            "the request query matches the plugin's claimed root",
        );
        checks.check(
            !plugin.claims(&format!("{root}_other/gain")),
            "a plugin whose root shares a prefix with this one is not claimed",
        );
    }

    println!("field read");
    {
        let response = plugin
            .call(&read(&format!("{root}/gain"), 1))?
            .expect("a non-notify read produces a response");
        let message = Message::from_slice(&response).expect("the response is a REPE frame");
        checks.check(
            message.json_body::<f64>().is_ok_and(|gain| gain == 1.0),
            "read /gain returns the constructed value",
        );
        checks.check(message.header.id == 1, "the response echoes the request id");
    }

    println!("field write");
    {
        let acknowledged = plugin.call(&write(&format!("{root}/channel"), 2, &6u32))?;
        checks.check(acknowledged.is_some(), "a write is acknowledged");

        let response = plugin
            .call(&read(&format!("{root}/channel"), 3))?
            .expect("a non-notify read produces a response");
        let message = Message::from_slice(&response).expect("the response is a REPE frame");
        checks.check(
            message.json_body::<u32>().is_ok_and(|channel| channel == 6),
            "the read observes what the write left behind",
        );
    }

    println!("method call");
    {
        let response = plugin
            .call(&write(&format!("{root}/calibrate"), 4, &8.0f64))?
            .expect("a non-notify call produces a response");
        let message = Message::from_slice(&response).expect("the response is a REPE frame");
        checks.check(
            message.json_body::<f64>().is_ok_and(|gain| gain == 4.0),
            "the method returns its computed result",
        );
    }

    println!("zero-argument method");
    {
        let response = plugin
            .call(&read(&format!("{root}/identify"), 5))?
            .expect("a non-notify call produces a response");
        let message = Message::from_slice(&response).expect("the response is a REPE frame");
        checks.check(
            message
                .json_body::<String>()
                .is_ok_and(|identity| identity.starts_with("instrument fw")),
            "the method result decodes as a string",
        );
    }

    println!("notify");
    {
        let response = plugin.call(&notify(&format!("{root}/reset"), 6))?;
        checks.check(
            response.is_none(),
            "a notify answers with size 0, which the host reads as `send nothing`",
        );
    }

    println!("unknown method");
    {
        let response = plugin
            .call(&read(&format!("{root}/absent"), 7))?
            .expect("an unknown method is answered, not dropped");
        let message = Message::from_slice(&response).expect("the response is a REPE frame");
        checks.check(
            message.error_code() == Some(ErrorCode::MethodNotFound),
            "an unknown method is method_not_found, not a dropped frame",
        );
    }

    println!("malformed frame");
    {
        // Not a REPE frame at all. The plugin owes an answer anyway: a host that
        // received nothing would have no way to tell a broken plugin from a
        // notify.
        let response = plugin
            .call(b"not a repe frame at all")?
            .expect("a malformed frame is answered, not silently dropped");
        let message = Message::from_slice(&response).expect("the answer is still a REPE frame");
        checks.check(
            message.is_error(),
            "a malformed frame is answered with an error",
        );
        checks.check(
            message.header.id == 0,
            "the id is 0, because it could not be read",
        );
    }

    println!("mounted on a router");
    {
        // The other way to host a plugin: hand it to `Router::with_fallback` and
        // let the router resolve its own routes first. `Plugin` implements
        // `HandlerErased`, so the mount is one line and the frame marshalling is
        // the crate's rather than the application's.
        //
        // A second `Plugin::load` of the same path reaches the same instance —
        // `dlopen` refcounts by path and nothing is ever unloaded — so this
        // shares state with the handle above rather than starting a new plugin.
        //
        // SAFETY: as above; the library is already resident.
        let mounted = unsafe { Plugin::load(path) }?;
        // And it says so. A host that reloads a path it was handed reports this
        // rather than a success, since nothing was read and nothing changed.
        checks.check(
            mounted.load_origin() == LoadOrigin::AlreadyResident,
            "a second load of one path reports itself as already resident",
        );
        // `_blocking`: a plugin call enters a library this process did not
        // build, for a time nothing here bounds, so on the WebSocket server it
        // belongs off the reader task. On the TCP servers the two are the same.
        let router = Router::new()
            .with_typed("/host/ping", |_| Ok(serde_json::json!("pong")))
            .with_fallback_blocking(std::sync::Arc::new(mounted));

        let frame = router
            .call(&write("/host/ping", 8, &serde_json::Value::Null))
            .expect("the host route answers");
        let message = Message::from_slice(&frame).expect("the response is a REPE frame");
        checks.check(
            message
                .json_body::<String>()
                .is_ok_and(|pong| pong == "pong"),
            "a route the host owns is not shadowed by the fallback",
        );

        let frame = router
            .call(&read(&format!("{root}/channel"), 9))
            .expect("the plugin answers through the router");
        let message = Message::from_slice(&frame).expect("the response is a REPE frame");
        checks.check(
            message.json_body::<u32>().is_ok_and(|channel| channel == 6),
            "a request under the plugin's root reaches the plugin",
        );
        checks.check(message.header.id == 9, "the mounted response keeps the id");

        let frame = router
            .call(&read("/not/claimed/by/anyone", 10))
            .expect("a declined path is answered rather than dropped");
        let message = Message::from_slice(&frame).expect("the response is a REPE frame");
        checks.check(
            message.error_code() == Some(ErrorCode::MethodNotFound),
            "a path outside the plugin's root is method_not_found",
        );

        // A notify must reach the plugin, not merely produce no response: the
        // dispatch layer discards a notify's response whatever the handler did,
        // so `is_none()` on its own would pass even if the plugin were skipped.
        // `calibrate` moves gain off its reset value; the notify puts it back.
        router.call(&write(&format!("{root}/calibrate"), 11, &8.0f64));
        checks.check(
            router.call(&notify(&format!("{root}/reset"), 12)).is_none(),
            "notify semantics carry through the mount",
        );
        let frame = router
            .call(&read(&format!("{root}/gain"), 13))
            .expect("answered");
        let message = Message::from_slice(&frame).expect("the response is a REPE frame");
        checks.check(
            message.json_body::<f64>().is_ok_and(|gain| gain == 1.0),
            "the notify reached the plugin and reset its gain",
        );
    }

    println!("shutdown");
    {
        // Last, because it is visible to everything above it. The handle is
        // consumed, so the borrow checker enforces what `plugin.h` states: no
        // further calls are made after this.
        //
        // A second `Plugin::load` of this same path would reach this same
        // shut-down instance — the library is never unloaded and `dlopen`
        // refcounts by path — so a host that shuts a plugin down has retired it
        // for the life of the process.
        plugin.shutdown();
    }

    Ok(checks.failures)
}

fn main() -> ExitCode {
    let Some(path) = std::env::args().nth(1) else {
        eprintln!("usage: plugin_host <path-to-plugin-library>");
        return ExitCode::from(2);
    };

    match run(&path) {
        Ok(0) => {
            println!("\nplugin host: PASS");
            ExitCode::SUCCESS
        }
        Ok(failures) => {
            println!("\nplugin host: FAIL ({failures} check(s))");
            ExitCode::FAILURE
        }
        Err(error) => {
            eprintln!("plugin host: {error}");
            ExitCode::from(2)
        }
    }
}
